src/utils/db/cid_index.rs

```diff
@@ -0,0 +1,79 @@
+use crate::utils::cid::CidCborExt;
+use cid::Cid;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::BTreeMap;
+use std::hash::Hasher;
+
+fn hash_cid(cid: Cid) -> usize {
+    usize::from_le_bytes(cid.hash().digest()[0..8].try_into().unwrap())
+}
+
+#[derive(Debug)]
+pub struct ProbingHashtableBuilder {
+    table: Vec<Option<(usize, u64)>>,
+}
```

Comment:
This looks a bit in-depth, and probably very difficult to review - why not something like
```rust
#[repr(transparent)]
struct CidHash(Cid);
impl std::hash::Hash for CidHash {
    fn hash<H>(&self, state: &mut H)
       where H: Hasher {
       state.write(todo!("get the bytes of randomness you want"));
    }
}
```

Reply:
To be clear, I strongly discourage writing a custom hashtable implementation

Reply:
I wouldn't be comfortable reviewing a custom hashtable implementation - can you think of other ways to achieve your goals of low memory usage for big indexes?

Could a custom `Hash` implementation help?
```
#[repr(transparent)]
struct CidHash(Cid);
impl std::hash::Hash for CidHash {
    fn hash<H>(&self, state: &mut H) where H: Hasher {
       state.write(todo!("get the bytes of randomness you want"));
    }
}
```

Or how about doing your own probing like this?
```rust
type QuickCidHashMap = HashMap<u64, Vec<(Cid, Location)>>;
```

And so your wire format is still pretty small?